src/xrCore/xrDebug.cpp

```diff
@@ -70,57 +77,204 @@ xrDebug::CrashHandler xrDebug::OnCrash = nullptr;
-        context.Ebp = (DWORD)&ebp;
-        context.Esp = (DWORD)&context;
+	if (!m_SymEngineInitialized)
+	{
+		DWORD dwOptions = SymGetOptions();
+		SymSetOptions(dwOptions | SYMOPT_DEFERRED_LOADS | SYMOPT_LOAD_LINES | SYMOPT_UNDNAME);
+
+		if (SymInitialize(GetCurrentProcess(), nullptr, TRUE))
+		{
+			m_SymEngineInitialized = true;
+		}
+	}
+
+	return m_SymEngineInitialized;
+}
+
+void xrDebug::DeinitializeSymbolEngine(void)
+{
+	if (m_SymEngineInitialized)
+	{
+		SymCleanup(GetCurrentProcess());
+
+		m_SymEngineInitialized = false;
+	}
+}
+
+xr_vector<xr_string> xrDebug::BuildStackTrace(PCONTEXT threadCtx, u16 maxFramesCount)
+{
+	SStringVec traceResult;
+	STACKFRAME stackFrame = { 0 };
+	xr_string frameStr;
+
+	if (!InitializeSymbolEngine())
+	{
+		Msg("[xrDebug::BuildStackTrace]InitializeSymbolEngine failed with error: %d", GetLastError());
+		return traceResult;
+	}
+
+	traceResult.reserve(maxFramesCount);
+
+#if defined XR_X64
+	stackFrame.AddrPC.Mode = AddrModeFlat;
+	stackFrame.AddrPC.Offset = threadCtx->Rip;
+	stackFrame.AddrStack.Mode = AddrModeFlat;
+	stackFrame.AddrStack.Offset = threadCtx->Rsp;
+	stackFrame.AddrFrame.Mode = AddrModeFlat;
+	stackFrame.AddrFrame.Offset = threadCtx->Rbp;
+#elif defined XR_X86	
+	stackFrame.AddrPC.Mode = AddrModeFlat;
+	stackFrame.AddrPC.Offset = threadCtx->Eip;
+	stackFrame.AddrStack.Mode = AddrModeFlat;
+	stackFrame.AddrStack.Offset = threadCtx->Esp;
+	stackFrame.AddrFrame.Mode = AddrModeFlat;
+	stackFrame.AddrFrame.Offset = threadCtx->Ebp;
+#else
+#	error CPU architecture is not supported.
 #endif
-        ex_ptrs.ContextRecord = &context;
-        ex_ptrs.ExceptionRecord = 0;
-        return BuildStackTrace(&ex_ptrs, buffer, capacity, lineCapacity);
-    }
-    return 0;
+
+	while (
+		GetNextStackFrameSring(&stackFrame, threadCtx, frameStr)
+		&&
+		traceResult.size() <= maxFramesCount
+		)
+	{
+		traceResult.push_back(frameStr);
+	}
+
+	DeinitializeSymbolEngine();
+
+	return traceResult;
 }
 
-size_t xrDebug::BuildStackTrace(EXCEPTION_POINTERS* exPtrs, char* buffer, size_t capacity, size_t lineCapacity)
+SStringVec xrDebug::BuildStackTrace(u16 maxFramesCount)
+{	
+	CONTEXT currentThreadCtx = { 0 };	
+
+	RtlCaptureContext(&currentThreadCtx);	/// GetThreadContext cann't be used on the current thread 
+	currentThreadCtx.ContextFlags = CONTEXT_FULL;
+
+	return BuildStackTrace(&currentThreadCtx, maxFramesCount);
+}
+
+void xrDebug::LogStackTrace(const char* header)
 {
-    memset(buffer, 0, capacity*lineCapacity);
-    auto flags = GSTSO_MODULE | GSTSO_SYMBOL | GSTSO_SRCLINE;
-    auto traceDump = GetFirstStackTraceString(flags, exPtrs);
-    int frameCount = 0;
-    while (traceDump)
-    {
-        lstrcpy(buffer + frameCount * lineCapacity, traceDump);
-        frameCount++;
-        traceDump = GetNextStackTraceString(flags, exPtrs);
-    }
-    return frameCount;
+	SStringVec stackTrace = BuildStackTrace();
+	Msg("%s", header);
+	for each (const auto& frame in stackTrace)
```

Comment:
Use standard range-based for instead of non-standard MS-specific 'for each'.

Reply:
Accept. I'll fix it